dec-no-dos/xorDecryptor.cpp

```diff
@@ -0,0 +1,168 @@
+#include "xorDecryptor.h"
+
+namespace xorDecryptor
+{
+	unsigned int XorDecryptor::standardKeyHash(int start, int end){
+		if (start == 0) {
+			return hash[end];
+		}
+
+		return hash[end] - hash[start - 1] * prime[end - start + 1];
+	}
+
+	XorDecryptor::XorDecryptor(std::vector<char> encryptedText, Parameters parameters):
+		validator(parameters)
+	{
+		this->encryptedText = encryptedText;
+		this->parameters = parameters;
+
+		hash.push_back(encryptedText[0]);
+		int bytesToHash = std::min(parameters.bytesToSearchForKeyIn, 
+			static_cast<int>(encryptedText.size()));
+		for (int i = 1; i < bytesToHash; i++) {
+			hash.push_back(hash[i - 1] * parameters.primeNumberForHashing 
+				+ encryptedText[i]);
+		}
+
+		prime.push_back(1);
+		for (int i = 1; i < bytesToHash; i++) {
+			prime.push_back(prime[i - 1] * parameters.primeNumberForHashing);
+		}
+
+	}
+
+	std::pair<bool, std::vector<char> > XorDecryptor::findKey()
+	{
+		PotentialKey bestKey;
+		bestKey.ratio = -1;
+		bool valid = false;
+
+		for (int i = 0; i < parameters.maxKeyLength; i++) {
+			PotentialKey newKey = findBestKeyForSetLength(i);
+			if (newKey.ratio < bestKey.ratio) {
+				bestKey = newKey;
+				valid = true;
+			}
+		}
+
+		return std::make_pair(valid, bestKey.key);
+	}
+
+	unsigned int XorDecryptor::rotatedKeyHash(int start, int rotation, int keySize){
+		if (rotation != 0) {

```

Comment:
So, basically, first part is the same.
`start + keySize - rotation - 1 = start + keySize - 0 - 1 = start + keySize - 1` -> the same as in line 57
`*prime[rotation] = *prime[0] = *1` -> prime^0th power = 1 (line 27).

If rotation is non-zero, you only add a second expression. So maybe rewrite it as:

```
unsigned int hash = standardKeyHash(start, start + keySize - rotation - 1) * prime[rotation];

if (rotation)
    hash += standardKeyHash(start - rotation, start - 1);

return hash;
```